### DISCUSSION TITLE
Media Library handlers when adding/removing files

### DISCUSSION THREAD

**Discussion by momentum-tunis** [OrchardCMS/OrchardCore#11590] [Q&A] [answered]
Hi,

Is it possible to catch adding files events in the media library ? I think it is possible with content items using IContentHandler, but I didn't find similar interfaces for media files.

Thanks!

**Comment by momentum-tunis** [comment, marked as answer]
There is nothing similar to ContentHandler for media files, I finally just used a middleware for this case.

**Reply by hishamco** [reply]
Can you show what you did for a reference?

**Reply by momentum-tunis** [reply]
I needed to add media files content to the Lucene index (pdf/doc/xls). With the new release 1.3, there is a new feature to incude pdf contents, but it allows only to return content items that reference pdf files (and not the files)

So I used the StoreDocumentsAsync of the LuceneIndexManager to add the content of files to the index. When an index is rebuilt from the admin section, we need to insert again the content of the media files. We should also update the index when adding/deleting/moving files, so I catch these action in a middleware I named OrchardExtenderMiddleware.

- Register the middleware in the Configure() method of Startup.cs:
app.UseMiddleware<OrchardExtenderMiddleware>();

- The content of OrchardExtenderMiddleware.cs:
```
public class OrchardExtenderMiddleware
{
	private readonly RequestDelegate _next;

	public OrchardExtenderMiddleware(RequestDelegate next)
	{
		_next = next;
	}

	public async Task Invoke(HttpContext context)
	{
		await _next(context);

		if (context.Request.Path.Value.ToLower().Contains("/lucene/rebuild/search"))
		{
			// TODO: index was rebuilt => append pdf/docs contents to index
		}

		if (context.Request.Path.Value.ToLower().Contains("/media/upload") &&
			context.Request.Form?.Files?.Count > 0)
		{
			// TODO: New files added, check if it'is an indexed media and add to index
		}

		if (context.Request.Path.Value.ToLower().Contains("/media/deletemedia"))
		{
			// QueryString = "?path=filepath"
			string path = context.Request.QueryString.Value.Substring(6);
			// TODO: file deleted, if it'is an indexed media file we need to remove it from the index
		}

		if (context.Request.Path.Value.ToLower().Contains("/media/movemedia"))
		{
			// TODO: file moved, if it'is an indexed media file we need to update the index
			// Orchard Controller method : MoveMedia(string oldPath, string newPath)
		}
	}
}

```

**Comment by hishamco** [comment]
Can we add  `IMediaHandler`?

**Reply by momentum-tunis** [reply]
It will be useful in many use cases, and for sure it will be needed if we want to add search feature for files. I could try to impelement that when I have time
